### PR TITLE
Let Gaussian accept LinearOperator represented sqrtprec

### LINKED ISSUE #419: Let Gaussian accept LinearOperator as sqrtprec

### PR DESCRIPTION
**Pull request by chaozg**
fixed #419 

Issue: it might open some opportunities if we allow a Gaussian distribution to be initialized with `scipy.sparse.linalg.LinearOperator`. 

What I did:
- update `get_sqrtprec_from_sqrtcov` in `cuqi/distribution/_gaussian.py` to accept a`LinearOperator` object as `sqrtprec`;
- add `check_if_conditional_from_attr` in `cuqi/utilities/_utilities.py`, following @nabriis 's comment below;
- `check_if_conditional_from_attr` is used in `get_indirect_variables` so that a `Gaussian` distribution initialized with `LinearOperator` is not detected as a conditional distribution (just a reminder: a LinearOperator is callable);
- add unit test `test_Gaussian_from_linear_operator_sqrtprec` in `tests/test_distribution.py`

Some notes (please correct me if I'm wrong here):
- For MCMC methods in general, we don't need `logdet` as it will be cancelled when evaluating the density function of the proposal and the current state;
- In CUQIpy, we need `logdet` defined to compute `logpdf`: I think this is somehow unnecessary and the unnormalized version `_logupdf` should be enough in place of `logpdf`. But I have to admit that I'm _not confident_ if this point still holds when it comes to Gibbs sampling. (This issue turned out to be an echo from the old days #16, as Nicolai pointed in the comments below)
- In order to make a Gaussian distribution initialized with `LinearOperator` work with the current samplers in CUQIpy except ROTs, we need to set `logdet` to some value. It doesn't matter to what value it is set as it will be cancelled.
- For RTOs, we don't evaluate `logpdf`, so we don't need to set `logdet`.

### CODE REVIEW

File: cuqi/distribution/_gaussian.py [hunk 1]
```diff
@@ -673,6 +673,13 @@ def get_sqrtprec_from_sqrtprec(dim, sqrtprec, sparse_flag):
         logdet = np.sum(-np.log(precision))
         rank = dim
 
+    # sqrtprec is LinearOperator (only for RTOs)
+    elif isinstance(sqrtprec, spa.linalg.LinearOperator):
```
Comment (nabriis) [inline cuqi/distribution/_gaussian.py:677]: I think this is a very reasonable addition. I would suggest adding to the existing unit tests cases for LinearOperator that check the logpdf values.

File: cuqi/distribution/_gaussian.py [hunk 2]
```diff
@@ -673,6 +673,13 @@ def get_sqrtprec_from_sqrtprec(dim, sqrtprec, sparse_flag):
...
+        sqrtprec = sqrtprec
+        if hasattr(sqrtprec, 'logdet'):
```
Comment (nabriis) [inline cuqi/distribution/_gaussian.py:679]: If logdet is not defined, what happens?

File: cuqi/utilities/_utilities.py [hunk 3]
```diff
@@ -64,7 +64,7 @@ def get_indirect_variables(dist):
     attributes = []
     for attribute in dist.get_mutable_variables():
         value = getattr(dist, attribute)
-        if callable(value):
+        if callable(value) and not isinstance(value, spslinalg.LinearOperator):
```
Comment (nabriis) [inline cuqi/utilities/_utilities.py:67]: First "callable" object that we want to interpret as a fix variable. I suggest we add a method to check if "value" is one of the allowed cases that are callable but still act as a well-defined ndarray

**Review by nabriis — CHANGES_REQUESTED** [review]
Hi @chaozg. I think it is a good idea to support LinearOperator. I had some sugggestions to add unit tests and also make the check in utilities more explicit by defining a method.

File: cuqi/distribution/_gaussian.py (hunk 1 above)
Comment (chaozg) [reply cuqi/distribution/_gaussian.py:677]: Done

File: cuqi/utilities/_utilities.py (hunk 3 above)
Comment (chaozg) [reply cuqi/utilities/_utilities.py:67]: Added a method `check_if_conditional_from_attr` in `cuqi/utilities/_utilities.py` for your suggested job

File: cuqi/distribution/_gaussian.py (hunk 2 above)
Comment (chaozg) [reply cuqi/distribution/_gaussian.py]: It's really the user's choice/responsibility if they give `logdet` or not. If they give `logdet`, we will record it and then `logpdf` can be computed; if not, `logpdf` can't be computed. 

For large matrices, it's not easy to compute `logdet`, but luckily we don't need it for samplers like `LinearRTO`

**Comment by chaozg** [comment]
Thanks @nabriis for your review and suggestions. I've updated the code to address your comments and am now requesting your further review.

File: cuqi/distribution/_gaussian.py (hunk 2 above)
Comment (nabriis) [reply cuqi/distribution/_gaussian.py]: This is OK for now.  I will create an issue related to the unormalized logpdf

File: cuqi/utilities/_utilities.py (hunk 3 above)
Comment (nabriis) [reply cuqi/utilities/_utilities.py]: LGTM

**Review by nabriis — APPROVED** [review]
Thanks @chaozg. LGTM

**Comment by chaozg** [comment]
Thanks @nabriis , I'm now requesting @jakobsj 's review.

File: cuqi/distribution/_gaussian.py (hunk 2 above)
Comment (nabriis) [reply cuqi/distribution/_gaussian.py]: There is an old issue related to this. See #16

Comment (chaozg) [reply cuqi/distribution/_gaussian.py]: > There is an old issue related to this. See #16

Hahaha, it turned out to be an echo from two years ago : )

**Review by jakobsj — APPROVED** [review]
Nice addition and the workaround for logdet seems fine. I added a small question, you can merge after addressing this.

File: cuqi/distribution/_gaussian.py [hunk 4]
```diff
@@ -666,6 +666,15 @@ def get_sqrtprec_from_sqrtprec(dim, sqrtprec, sparse_flag):
         logdet = np.sum(-np.log(sqrtprec.data**2))
         rank = dim
 
+    # sqrtprec is LinearOperator
+    elif isinstance(sqrtprec, spa.linalg.LinearOperator):
+        sqrtprec = sqrtprec
```
[the remark this answers is not in the document]
Comment (chaozg) [reply cuqi/distribution/_gaussian.py]: Thanks for the spot. It's indeed unnecessary.

**Pull request merged**